### DISCUSSION TITLE
How would I reuse definitions from objects?

### DISCUSSION THREAD

**Discussion by digital-h** [victools/jsonschema-generator#301] [Q&A] [answered]
Hi,

imagine I have the following 2 classes in one package

```
package myPackage;

public class Foo {
   public String baz;
}
```
```
package myPackage;

public class Bar {
   public Foo foo;
}
```

and I run the maven-plugin defining the package

```
<configuration>
    <packageNames>
        <packageName>myPackage</packageName>
    </packageNames>
</configuration>
```

What currently happens, that two seperate schema are generated for Foo and Bar. But in Bar everything is generated for Foo as well instead of referencing to the Foo-Schema. Is it possible to configure this in a way?

Instead for Bar.schema as it is generated right now
```
{
  "$schema" : "http://json-schema.org/draft-07/schema#",
  "type" : "object",
  "properties" : {
    "foo" : {
      "type" : "object",
      "properties" : {
        "baz" : {
          "type" : "string"
        }
      },
      "additionalProperties" : false
    }
  },
  "additionalProperties" : false
}
```

it should become:
```
{
  "$schema" : "http://json-schema.org/draft-07/schema#",
  "type" : "object",
  "properties" : {
    "foo" : {
      "$ref": "Foo.schema.json",
    }
  },
  "additionalProperties" : false
}
```
Thanks for your help

**Comment by CarstenWickner** [comment, marked as answer]
Hi @digital-h,

Fortunately, I already answered a very similar question here: **https://github.com/victools/jsonschema-generator/issues/246#issuecomment-1099666190**
That includes a complete code example.

----

More generically speaking: What you ask for requires some custom configuration, as per this FAQ entry:
https://victools.github.io/jsonschema-generator/#how-to-reference-a-separate-schema-file
> ### How to reference a separate schema/file?
> By using `withCustomDefinitionProvider()` – one of the [advanced configurations](https://victools.github.io/jsonschema-generator/#generator-advanced-configurations) – you can fully control the contents of a type's sub-schema. Simply create a node that only contains your custom/external reference instead of the actual schema. It is recommended to mark the custom definition as "to be inlined", in order to avoid an extra entry in the "definitions"/"$defs".
```java
configBuilder.forTypesInGeneral()
        .withCustomDefinitionProvider((javaType, context) -> {
            if (javaType.getErasedType() != MyExternalType.class) {
                // other types should be treated normally
                return null;
            }
            // define your custom reference value
            String refValue = "./" + javaType.getErasedType().getSimpleName();
            // produce the sub-schema that only contains your custom reference
            ObjectNode customNode = context.getGeneratorConfig().createObjectNode()
                    .put(context.getKeyword(SchemaKeyword.TAG_REF), refValue);
            return new CustomDefinition(customNode,
                    // avoid the creation of a reference to your custom reference schema
                    CustomDefinition.DefinitionType.INLINE,
                    // still allow for collected schema attributes to be added
                    CustomDefinition.AttributeInclusion.YES);
        });
```

If you want this kind of custom configuration to be applied by the Maven plugin, you'd have to wrap it into a `Module` and then reference it in your plugin configuration as mentioned here:
https://victools.github.io/jsonschema-generator/#further-configurations-through-modules
> You can also group any kind of configuration into a Module of your own and include it via its full class name.
> Make sure your custom module is on the classpath (considering the project itself as well as all compile and runtime dependencies) and has a default constructor.
```xml
<modules>
    <module>
        <className>com.myOrg.myApp.CustomModule</className>
    </module>
</modules>
```
